x/delayedack/ibc_middleware.go

```diff
@@ -1,323 +1,214 @@
-		return err
+	if !w.Keeper.IsRollappsEnabled(ctx) {
+		return w.IBCModule.OnTimeoutPacket(ctx, packet, relayer)
 	}
 
-	proofHeight, err := im.GetProofHeight(ctx, commontypes.RollappPacket_ON_TIMEOUT, rollappPortOnHub, rollappChannelOnHub, packet.Sequence)
+	transfer, err := w.GetValidTransferWithFinalizationInfo(ctx, packet, commontypes.RollappPacket_ON_TIMEOUT)
 	if err != nil {
-		logger.Error("Failed to get proof height from packet", "err", err)
-		return err
-	}
-	finalized, err := im.CheckIfFinalized(ctx, rollappID, proofHeight)
-	if err != nil {
-		logger.Error("Failed to check if packet is finalized", "err", err)
+		l.Error("Get valid rollapp and transfer.", "err", err)
 		return err
 	}
 
-	if finalized {
-		logger.Debug("Skipping IBC transfer OnTimeoutPacket as the packet proof height is already finalized")
-		return im.IBCModule.OnTimeoutPacket(ctx, packet, relayer)
+	if !transfer.IsRollapp() || transfer.Finalized {
+		return w.IBCModule.OnTimeoutPacket(ctx, packet, relayer)
 	}
 
 	// Run the underlying app's OnTimeoutPacket callback
 	// with cache context to avoid state changes and report the timeout result.
 	// Only save the packet if the underlying app's callback succeeds.
 	cacheCtx, _ := ctx.CacheContext()
-	err = im.IBCModule.OnTimeoutPacket(cacheCtx, packet, relayer)
+	err = w.IBCModule.OnTimeoutPacket(cacheCtx, packet, relayer)
 	if err != nil {
 		return err
 	}
-	// Save the packet data to the store for later processing
-	rollappPacket := commontypes.RollappPacket{
-		RollappId:   rollappID,
-		Packet:      &packet,
-		Status:      commontypes.Status_PENDING,
-		Relayer:     relayer,
-		ProofHeight: proofHeight,
-		Type:        commontypes.RollappPacket_ON_TIMEOUT,
-	}
-	im.SetRollappPacket(ctx, rollappPacket)
-
-	logger.Debug("Set rollapp packet",
-		"rollappID", rollappPacket.RollappId,
-		"proofHeight", rollappPacket.ProofHeight,
-		"type", rollappPacket.Type)
 
-	err = im.EIBCDemandOrderHandler(ctx, rollappPacket, *transferPacketData)
-	if err != nil {
-		return err
-	}
+	rollappPacket := w.getSavedPacket(ctx, l, packet, transfer, relayer, commontypes.RollappPacket_ON_TIMEOUT, nil)
 
-	return nil
+	return w.EIBCDemandOrderHandler(ctx, rollappPacket, transfer.FungibleTokenPacketData)
 }
 
-// GetProofHeight returns the proof height of the packet
-func (im IBCMiddleware) GetProofHeight(ctx sdk.Context, packetType commontypes.RollappPacket_Type,
-	rollappPortOnHub string, rollappChannelOnHub string, sequence uint64,
-) (uint64, error) {
-	packetId := commontypes.NewPacketUID(packetType, rollappPortOnHub, rollappChannelOnHub, sequence)
-	height, ok := types.FromIBCProofContext(ctx, packetId)
-	if ok {
-		return height.RevisionHeight, nil
-	} else {
-		err := errors.New("failed to get proof height from context")
-		ctx.Logger().Error(err.Error(), "packetId", packetId)
-		return 0, err
+// savePacket the packet to the store for later processing and returns it
```

Comment:
`get` is misleading as this func creates and Sets new rollappPacket to the store.
also wouldn't log inside this function. bit messy

Reply:
renamed to savePacket, deleted log